custom_components/control4/fan.py

```diff
@@ -0,0 +1,179 @@
+                item_name = str(item["name"])
+                item_id = item["id"]
+                item_area = item["roomName"]
+                item_parent_id = item["parentId"]
+
+                item_manufacturer = None
+                item_device_name = None
+                item_model = None
+
+                for parent_item in items_of_category:
+                    if parent_item["id"] == item_parent_id:
+                        item_manufacturer = parent_item["manufacturer"]
+                        item_device_name = parent_item["name"]
+                        item_model = parent_item["model"]
+            else:
+                continue
+        except KeyError:
+            _LOGGER.exception(
+                "Unknown device properties received from Control4: %s",
+                item,
+            )
+            continue
+
+
+        item_attributes = await director_get_entry_variables(hass, entry, item_id)
+
+        entity_list.append(
+            Control4Fan(
+                entry_data,
+                entry,
+                item_name,
+                item_id,
+                item_device_name,
+                item_manufacturer,
+                item_model,
+                item_parent_id,
+                item_area,
+                item_attributes,
+            )
+        )
+
+    async_add_entities(entity_list, True)
+
+
+class Control4Fan(Control4Entity, FanEntity):
+    """Control4 fan entity."""
+
+    def create_api_object(self):
+        """Create a pyControl4 device object.
+           This exists so the director token used is always the latest one, without needing to re-init the entire entity.
+        """
+        return C4Fan(self.entry_data[CONF_DIRECTOR], self._idx)
+    
+    async def _update_callback(self, device, message):
+        """Update state attributes in hass after receiving a Websocket update for our item id/parent device id."""
+        # Message will be False when a Websocket disconnect is detected
+        if message is False:
+            self._attr_available = False
+        elif message["evtName"] == "OnDataToUI":
+            self._attr_available = True
+            data = message["data"]
+            if "fan_state" in data:
+                fan_state = data["fan_state"]
+                if "current_speed" in fan_state:
+                    _LOGGER.debug("Zupdate fan speed %s", str(fan_state))
+                    current_speed = fan_state.pop("current_speed")
+                    self._extra_state_attributes["CURRENT_SPEED"] = current_speed
+            if "fan_setup" in data:
+                fan_setup = data["fan_setup"]
+                if "preset_speed" in fan_setup:
+                    preset_speed = fan_setup.pop("preset_speed")
+                    self._extra_state_attributes["PRESET_SPEED"] = preset_speed
+
+        self.schedule_update_ha_state()
+
+
+    @property
+    def percentage_step(self) -> float:
+        """Return the step size for percentage."""
+        return 25
```

Comment:
Do all fans have a 25% step size?

Reply:
Probably not, all the control4 fan controllers I have have 4 preset modes.  I will dig a little and see what more I can find out here.

Reply:
It would be ideal if we could determine this from the Control4 API response, if they provide that information in the device info